src/core/tsi/ssl_transport_security_utils.cc

```diff
@@ -247,4 +251,62 @@ tsi_result SslProtectorUnprotect(const unsigned char* protected_frames_bytes,
   return result;
 }
 
+bool VerifyCrlSignature(X509_CRL* crl, X509* issuer) {
+  if (issuer == nullptr || crl == nullptr) {
+    return false;
+  }
+  EVP_PKEY* ikey = X509_get_pubkey(issuer);
+  if (ikey == nullptr) {
+    // Can't verify signature because we couldn't get the pubkey, fail the
+    // check.
+    EVP_PKEY_free(ikey);
+    return false;
+  }
+  bool ret = X509_CRL_verify(crl, ikey) == 1;
+  EVP_PKEY_free(ikey);
+  return ret;
+}
+
+bool VerifyCrlCertIssuerNamesMatch(X509_CRL* crl, X509* cert) {
+  if (cert == nullptr || crl == nullptr) {
+    return false;
+  }
+  X509_NAME* cert_issuer_name = X509_get_issuer_name(cert);
+  if (cert == nullptr) {
+    return false;
+  }
+  X509_NAME* crl_issuer_name = X509_CRL_get_issuer(crl);
+  if (crl_issuer_name == nullptr) {
+    return false;
+  }
+  return X509_NAME_cmp(cert_issuer_name, crl_issuer_name) == 0;
+}
+
+bool HasCrlSignBit(X509* cert) {
+// X509_get_key_usage is not in 1.0.2. Just make this return true.
+#if OPENSSL_VERSION_NUMBER < 0x10100000
+  return true;
```

Comment:
If my memory serves me correctly and you must support old openssl, in 1.0.2 this was just exposed and didn't have an accessor, so you can just look at cert->ex_kusage and check the flag, rather than just giving up and allowing any certificate to be used to sign a CRL.

Reply:
Just want to double check a behavior with you
> X509_get_key_usage() returns the value of the key usage extension. If key usage is present will return zero or more of the flags: KU_DIGITAL_SIGNATURE, KU_NON_REPUDIATION, KU_KEY_ENCIPHERMENT, KU_DATA_ENCIPHERMENT, KU_KEY_AGREEMENT, KU_KEY_CERT_SIGN, KU_CRL_SIGN, KU_ENCIPHER_ONLY or KU_DECIPHER_ONLY corresponding to individual key usage bits. If key usage is absent then UINT32_MAX is returned.

UINT32_MAX will mean that every usage is seen as valid - I've done this for 1.0.2 in the case of a missing extension too, does that sound right to you

Reply:
yeah if there's no key usage present then effectively that's what you need to do.